lib/accesslists/walk.go

```diff
@@ -0,0 +1,305 @@
+/*
+ * Teleport
+ * Copyright (C) 2025  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package accesslists
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	accesslistv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/accesslist/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/accesslist"
+	"github.com/gravitational/teleport/api/utils/clientutils"
+)
+
+var (
+	// Not using trace errors is intentional here as those errors are expected.
+	// They are used to control the graph traversal. Using trace.Errorf() or
+	// other trace errors would cause useless stacktrace captures each time an
+	// error is returned.
+	skipLeg = errors.New("skip this access leg")
+	skipAll = errors.New("skip everything and stop the walk")
+)
+
+// walkFunc is the type of the function called by walk on each
+// access list member and nested access list part of the access graph.
+// The error result returned by the function controls how accessPath continues.
+// If the function returns the special value SkipDir, Walk skips the current directory (path if info.IsDir() is true, otherwise path's parent directory). If the function returns the special value SkipAll, Walk skips all remaining files and directories. Otherwise, if the function returns a non-nil error, Walk stops entirely and returns that error.
+type walkFunc func(path accessPath) error
+
+// walkUntilUser returns a walkFunc that filters out every invalid
+// accessLeg. Invalid access legs are:
+// - expired legs
+// - legs granting access to a different user
+// - legs granting access to a list whose requirements are not met by the user
+func walkUntilUser(user types.User, now time.Time) (walkFunc, func() (accesslistv1.AccessListUserAssignmentType, error)) {
+	skipped := make([]skippedAccessPath, 0)
```

Comment:
That's a total nitpick, but I think there is no point of initialising it here

```suggestion
	var skipped []skippedAccessPath
```